File: GymApp/composeApp/src/commonMain/kotlin/com/gym/gymapp/data/repository/MemberRepository.kt
package com.gym.gymapp.data.repository

import com.gym.gymapp.data.models.*
import com.gym.gymapp.network.ApiEndpoints
import com.gym.gymapp.network.NetworkClient
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.http.*
import io.ktor.http.content.*

class MemberRepository {
    private val client = NetworkClient.client
    
    suspend fun getMembers(gymId: String? = null): Result<List<Member>> {
        return try {
            val response: ApiResponse<List<Member>> = client.get(ApiEndpoints.MEMBERS) {
                if (gymId != null) {
                    parameter("gymId", gymId)
                }
            }.body()
            Result.success(response.data)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun createMember(request: CreateMemberRequest): Result<Member> {
        return try {
            val response = client.post(ApiEndpoints.MEMBERS) {
                contentType(ContentType.Application.Json)
                setBody(request)
            }
            if (response.status == HttpStatusCode.Created || response.status == HttpStatusCode.OK) {
                val apiRes: ApiResponse<Member> = response.body()
                Result.success(apiRes.data)
            } else {
                Result.failure(Exception("Failed to create member"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun uploadImage(imageBytes: ByteArray): Result<String> {
        return try {
            val response: io.ktor.client.statement.HttpResponse = client.submitFormWithBinaryData(
                url = "${ApiEndpoints.MEMBERS}/upload",
                formData = formData {
                    append("file", imageBytes, Headers.build {
                        append(HttpHeaders.ContentType, "image/jpeg")
                        append(HttpHeaders.ContentDisposition, "filename=\"profile.jpg\"")
                    })
                }
            )
            
            if (response.status == HttpStatusCode.OK || response.status == HttpStatusCode.Created) {
                val apiRes: ApiResponse<Map<String, String>> = response.body()
                Result.success(apiRes.data["url"] ?: "")
            } else {
                Result.failure(Exception("Image upload failed: ${response.status}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun updateMember(id: String, request: UpdateMemberRequest): Result<Member> {
        return try {
            val response = client.put("${ApiEndpoints.MEMBERS}/$id") {
                contentType(ContentType.Application.Json)
                setBody(request)
            }
            if (response.status == HttpStatusCode.OK) {
                val apiRes: ApiResponse<Member> = response.body()
                Result.success(apiRes.data)
            } else {
                Result.failure(Exception("Failed to update member"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun deleteMember(id: String): Result<Boolean> {
        return try {
            val response = client.delete("${ApiEndpoints.MEMBERS}/$id")
            if (response.status == HttpStatusCode.OK || response.status == HttpStatusCode.NoContent) {
                Result.success(true)
            } else {
                Result.failure(Exception("Failed to delete member"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: gym_flutter_app/lib/data/repositories/member_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../models/api_response.dart';
import '../models/member_model.dart';
import 'api_client.dart';

final memberRepositoryProvider = Provider<MemberRepository>((ref) {
  final dio = ref.watch(apiClientProvider);
  return MemberRepository(dio);
});

class MemberRepository {
  final Dio _dio;
  MemberRepository(this._dio);

  Future<ApiResponse<Member>> createMember(Map<String, dynamic> data) async {
    try {
      final response = await _dio.post('/members', data: data);
      return ApiResponse.fromJson(response.data, (data) => Member.fromJson(data));
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }

  Future<ApiResponse<String>> uploadImage(List<int> imageBytes) async {
    try {
      final formData = FormData.fromMap({
        'file': MultipartFile.fromBytes(imageBytes, filename: 'profile.jpg'),
      });
      final response = await _dio.post('/members/upload', data: formData);
      return ApiResponse.fromJson(response.data, (data) => (data as Map)['url'] ?? '');
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }

  Future<ApiResponse<List<Member>>> getMembers({required String gymId, int page = 1, int limit = 50}) async {
    try {
      final response = await _dio.get('/members', queryParameters: {'gymId': gymId, 'page': page, 'limit': limit});
      return ApiResponse.fromJson(response.data, (data) => (data as List).map((m) => Member.fromJson(m)).toList());
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }

  Future<ApiResponse<Member>> getMember(String id) async {
    try {
      final response = await _dio.get('/members/$id');
      return ApiResponse.fromJson(response.data, (data) => Member.fromJson(data));
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }

  Future<ApiResponse<Member>> updateMember(String id, Map<String, dynamic> data) async {
    try {
      final response = await _dio.patch('/members/$id', data: data);
      return ApiResponse.fromJson(response.data, (data) => Member.fromJson(data));
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }

  Future<ApiResponse<void>> deleteMember(String id) async {
    try {
      final response = await _dio.delete('/members/$id');
      return ApiResponse(status: response.data['status'] ?? false, message: response.data['message'] ?? '');
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }

  Future<ApiResponse<Member>> toggleMemberStatus(String id) async {
    try {
      final response = await _dio.post('/members/$id/toggle-status');
      return ApiResponse.fromJson(response.data, (data) => Member.fromJson(data));
    } catch (e) {
      return ApiResponse(status: false, message: e.toString());
    }
  }
}

